src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs

```diff
@@ -3507,18 +3459,17 @@ private void SerializeMarshallingDescriptor(IMarshallingInformation marshallingI
                     writer.WriteUInt16(0); // padding
 
                     object marshaller = marshallingInformation.GetCustomMarshaller(Context);
-                    ITypeReference marshallerTypeRef = marshaller as ITypeReference;
-                    if (marshallerTypeRef != null)
+                    switch (marshaller)
                     {
-                        this.SerializeTypeName(marshallerTypeRef, writer);
-                    }
-                    else if (marshaller != null)
-                    {
-                        writer.WriteSerializedString((string)marshaller);
-                    }
-                    else
-                    {
-                        writer.WriteByte(0);
+                        case ITypeReference marshallerTypeRef:
+                            this.SerializeTypeName(marshallerTypeRef, writer);
+                            break;
+                        case { }:
```

Comment:
not: `case string s`?

Reply:
I was trying not to modify the semantics of the code as written.

---
In reply to: [291407933](https://github.com/dotnet/roslyn/pull/36219#discussion_r291407933) [](ancestors = 291407933)